metronome/checkpointing/interpreter/src/io/iohk/metronome/checkpointing/interpreter/InterpreterService.scala

```diff
@@ -0,0 +1,161 @@
+package io.iohk.metronome.checkpointing.interpreter
+
+import cats.implicits._
+import cats.effect.{Concurrent, Timer, Resource}
+import io.iohk.metronome.checkpointing.CheckpointingAgreement
+import io.iohk.metronome.checkpointing.interpreter.messages.InterpreterMessage
+import io.iohk.metronome.checkpointing.interpreter.tracing.InterpreterEvent
+import io.iohk.metronome.checkpointing.models.Transaction
+import io.iohk.metronome.networking.LocalConnectionManager
+import io.iohk.metronome.tracer.Tracer
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+import io.iohk.metronome.core.messages.RPCPair
+
+/** The `InterpreterService` is to be used on the Interpreter side to
+  * manage the behind-the-scenes messaging with the Service.
+  */
```

Comment:
Again, I think this description could be clearer. I'm actually lost here. I thought this was meant to be used on the PoW side, but then it depends on the `InterpreterRPC`. Could you clarify?

Reply:
Is it actually meant for both the PoW node and Hotstuff node?

Reply:
Yes, I'm guilty as charged abusing the "Service" name, sorry for the confusion. This one is a service in the technical sense as it will actually handle to messages coming from the Checkpointing Service. Okay, you are right, I will add "Checkpointing" to disambiguate. Would be nice to find a different work for the "BFT + Checkpointing ~~Service~~ ???" side.

> I thought this was meant to be used on the PoW side, but then it depends on the `InterpreterRPC`. Could you clarify?

Yeah, so the idea is that the `InterpreterRPC` is an abstract interface, these are the RPC methods that the Interpreter implements. However, that doesn't say it lives on one place or the other: on the Interpreter side there should be a "server/service" that handles the requests, and on the Service side there would be a "client" that sends the requests. 

This is similar to [gRPC](https://cloud.google.com/apis/design/naming_convention#interface_names): we could have a `service Interpreter` with methods like `rpc CreateBlockBody` in the `checkpointing.proto` file, and then you can generate the server and the client side stubs as well. I'm aware that `service Service` would make a bad name if it had to be in gRPC :)

`InterpreterRPC` is passed to `InterpreterService.apply` because that's how for example the Mantis team would inject their logic into the machinery. The return value is a `ServiceRPC` because that's the interface _they_ can call to send messages to the Service.

On the [Service side](https://github.com/input-output-hk/metronome/pull/53), we have the opposite where the `ServiceRPC` implementation is passed to the `InterpreterClient` and an `InterpreterRPC` is returned. 

I can understand that he naming seems arbitrary, since both have client as well as server aspects, being a bi-directional setup. 

Do you have a better suggestion?

Reply:
> on the Interpreter side there should be a "server/service" that handles the requests, and on the Service side there would be a "client" that sends the requests.

Ok, I get it now :+1:

> Do you have a better suggestion?

Only to try to disambiguate by saying "Checkpointing Service". If you keep those words separate, then you're not really directly referring to the class

Reply:
Added "Checkpointing" throughout this class.